reward_calcs/reward_calcs.cpp

```diff
@@ -0,0 +1,474 @@
+                auto claim = static_cast<int64_t>(static_cast<fixp_t>(max_rewards) * static_cast<fixp_t>(FP(v->weight) / weights_sum));
+                
+                //eosio_assert(claim <= unclaimed_rewards, ("LOGIC ERROR! forum::pay_curators: claim(" + 
+                 //   std::to_string(claim) + ") > " + "unclaimed_rewards(" + std::to_string(unclaimed_rewards) + ")").c_str()); 
+                //^to_string isn't working for now: https://github.com/EOSIO/eosio.cdt/issues/95
+                eosio_assert(claim <= unclaimed_rewards, "LOGIC ERROR! forum::pay_curators: claim > unclaimed_rewards");
+                
+                if(claim > 0) {
+                    unclaimed_rewards -= claim;
+                    payto(v->voter, eosio::asset(claim, tokensymbol), static_cast<enum_t>(payment_t::VESTING));
+                }
+            }
+            v = idx.erase(v);
+        }
+        return unclaimed_rewards;
+    }
+    
+    auto get_pool(reward_pools& pools, uint64_t time) {        
+        eosio_assert(pools.begin() != pools.end(), "forum::get_pool: [pools] is empty");
+        
+        auto pool = pools.upper_bound(time);
+        
+        eosio_assert(pool != pools.begin(), "forum::get_pool: can't find an appropriate pool");        
+        return (--pool);
+    }
+    
+    static fixp_t get_delta(fixp_t old_val, fixp_t new_val, const funcinfo& func) {
+
+        atmsp::machine<fixp_t> mchn;
+        func.code.to_machine(mchn);
+        
+        mchn.var[0] = std::min(old_val, FP(func.maxarg)); 
+        fixp_t old_fn = mchn.run();        
+        mchn.var[0] = std::min(new_val, FP(func.maxarg));
+        fixp_t new_fn = mchn.run();
+        
+        return (new_fn - old_fn);
+    }
+    
+    static fixp_t calc_1d(atmsp::machine<fixp_t>& machine, fixp_t arg) {
+        machine.var[0] = arg;
+        return machine.run();
+    }
+   
+    static fixp_t positive_safe_cast(base_t arg) {
+        eosio_assert(arg > 0, "forum::positive_safe_cast: arg <= 0");
+        if (fixp_t::exponent < 0) //replace with [if constexpr] in the future
+            eosio_assert(arg <= static_cast<base_t>(std::numeric_limits<fixp_t>::max()), "forum::positive_safe_cast: arg > max possible value");
+        return static_cast<fixp_t>(arg);
+    }
+    
+    static fixp_t get_limit_prop(base_t arg) {
+        fixp_t ret = positive_safe_cast(arg);
+        ret /= static_cast<fixp_t>(ONE_HUNDRED_PERCENT);
+        return std::min(ret, static_cast<fixp_t>(1));
+    }
+   
+    static void check_positive_monotonic(atmsp::machine<fixp_t>& machine, fixp_t max_arg, const std::string& name, bool inc) {
+        fixp_t prev_res = calc_1d(machine, max_arg);
+        if(!inc)
+            eosio_assert(prev_res >= static_cast<fixp_t>(0), ("forum::check positive failed for " + name).c_str());
+        fixp_t cur_arg = max_arg;
+        for(size_t i = 0; i < CHECK_MONOTONIC_STEPS; i++) {
+            cur_arg /= static_cast<fixp_t>(2);           
```

Comment:
The logarithmic scale is better because we want to have more checks at points that are closer to zero.
Yes, netshares can be negative, but in this case the message doesn't get rewards (for now, but can have impact on pool's properties). I guess we should cutoff negative args, so I added this in calc_1d. Thx.